domain/network/state/space.go

```diff
@@ -0,0 +1,334 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3"
+
+	coreDB "github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/internal/database"
+)
+
+// State represents a type for interacting with the underlying state.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new State for interacting with the underlying state.
+func NewState(factory coreDB.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// AddSpace creates and returns a new space.
+func (st *State) AddSpace(
+	ctx context.Context,
+	uuid utils.UUID,
+	name string,
+	providerID network.Id,
+	subnetIDs []string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	insertSpaceStmt := `
+INSERT INTO space (uuid, name)
+VALUES (?, ?)`
+
+	insertProviderStmt := `
+INSERT INTO provider_space (provider_id, space_uuid)
+VALUES (?, ?)`
+
+	fanSubnetsBinds, fanSubnetsVals := database.SliceToPlaceholder(subnetIDs)
+	findFanSubnetsStmt := fmt.Sprintf(`
+SELECT subject_subnet_uuid
+FROM   subnet_association
+WHERE  association_type_id = 0 AND associated_subnet_uuid IN (%s)`, fanSubnetsBinds)
+
+	checkInputSubnetsStmt := fmt.Sprintf(`
+SELECT uuid
+FROM   subnet
+JOIN   subnet_type
+ON     subnet.subnet_type_id = subnet_type.id
+WHERE  subnet_type.is_space_settable = FALSE AND subnet.uuid IN (%s)`, fanSubnetsBinds)
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		if _, err := tx.ExecContext(ctx, insertSpaceStmt, uuid.String(), name); err != nil {
+			return errors.Annotatef(err, "inserting space uuid %q into space table", uuid.String())
+		}
+		if providerID != "" {
+			if _, err := tx.ExecContext(ctx, insertProviderStmt, providerID, uuid.String()); err != nil {
+				return errors.Annotatef(err, "inserting provider id %q into provider_space table", providerID)
+			}
+		}
+		// We must first check on the provided subnet ids to validate
+		// that none of them is a fan subnet.
+		fanSubnets, err := tx.QueryContext(ctx, checkInputSubnetsStmt, fanSubnetsVals...)
```

Comment:
You don't need to do this. The fact that they are Fan subnets is not the pertinent thing here. 

The pertinent thing is that they _happen_ to be types of subnets that cannot have space explicitly set. 

That fact is checked below with `checkInputSubnetsStmt`, which should be hoisted to the top of this method before any writes.

Reply:
So if I understood correctly, the query statement (`checkInputSubnetsStmt`) was correct (because it already retrieved all the subnets `IN` the inputs that have `is_space_settable = FALSE`). The naming was a bit misleading though. Also, I moved this entire check at the beginning of the tx.

This is done in 51238c1363, let me know if I misunderstood your comment

Reply:
also @manadart I left the error message as `cannot set space for FAN subnet UUIDs %q - it is always inherited from underlay` so it's easier to migrate tests, but if you want I  can rephrase it